### PR TITLE
[release/3.1] Add new System.Net.Http.Json project/namespace

### PR DESCRIPTION
**Pull request by jozkee**
Port of https://github.com/dotnet/runtime/pull/33459
**Note that the runtime PR is still open, so please add all feedback that is not related to release/3.1 there** (all new commits on that PR will be ported into here as well)
# Description
Serializing and deserializing JSON payloads from the network is a very common operation for clients, especially in the upcoming Blazor environment. Right now, sending a JSON payload to the server requires multiple lines of code, which will be a major speed bump for those customers. We'd like to add extension methods on top of HttpClient that allows doing those operations with a single method call.

This creates a package that targets netstandard2.0 only. 
This does not ship inbox in this branch.

# Customer Impact
This provides a solution that serves as a replacement of the old-fashioned [Microsoft.AspNet.WebApi.Client](https://www.nuget.org/packages/Microsoft.AspNet.WebApi.Client), that is compatible with netstandard2.0 and can be used by Blazor WebAssembly apps.

# Regression
None. 
This is a brand new project.

# Testing
Added unit tests to verify correct behavior of API surface.
Added unit tests to verify correct behavior of transcoding semantics.

# Risk
This is a whole new API that has not even shipped as a preview in runtime, and it is still on active development.

### CODE REVIEW

**Comment by joperezr** [comment]
cc: @ericstj @jeffhandley

**Comment by joperezr** [comment]
cc: @terrajobst

**Comment by ericstj** [comment]
@Jozkee the tests you've added are failing:
https://dev.azure.com/dnceng/public/_build/results?buildId=561879&view=ms.vss-test-web.build-test-results-tab&runId=17681388&resultId=112227&paneView=debug

**Comment by ericstj** [comment]
All the failures appear to be:
```
System.ArgumentException : The output byte buffer is too small to contain the encoded data, encoding 'Unicode (UTF-8)' fallback 'System.Text.EncoderReplacementFallback'.\r\nParameter name: bytes
```

**Comment by joperezr** [comment]
@ericstj yeah we actually knew before hand that those tests would fail, he was in the middle of working on fixing them when we started porting the branch to release/3.1, once he fixes all those in runtime repo we will port the fix back here. I'm looking into the packaging test failures, as they are related to the new package.

File: src/System.Net.Http.Json/ref/System.Net.Http.Json.cs [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+        public static System.Threading.Tasks.Task<T> ReadFromJsonAsync<T>(this System.Net.Http.HttpContent content, System.Text.Json.JsonSerializerOptions? options = null, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) { throw null; }
+    }
+    public partial class JsonContent : System.Net.Http.HttpContent
+    {
+        internal JsonContent() { }
+        public System.Type ObjectType { get { throw null; } }
+        public object? Value { get { throw null; } }
+        public static System.Net.Http.Json.JsonContent Create(object? inputValue, System.Type inputType, System.Net.Http.Headers.MediaTypeHeaderValue mediaType, System.Text.Json.JsonSerializerOptions? options = null) { throw null; }
+        public static System.Net.Http.Json.JsonContent Create<T>(T value, System.Net.Http.Headers.MediaTypeHeaderValue mediaType, System.Text.Json.JsonSerializerOptions? options = null) { throw null; }
```
Comment (stephentoub) [inline src/System.Net.Http.Json/ref/System.Net.Http.Json.cs]: Both here and above, mediaType is supposed to have `= null`, right?

Comment (stephentoub) [inline src/System.Net.Http.Json/ref/System.Net.Http.Json.cs]: We have `T value` and `object? inputValue`... should they have the same name?

File: src/System.Net.Http.Json/src/Properties/InternalsVisibleTo.cs [hunk 2]
```diff
@@ -0,0 +1,7 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime.CompilerServices;
+
+[assembly: InternalsVisibleTo("System.Net.Http.Json.Functional.Tests, PublicKey=00240000048000009400000006020000002400005253413100040000010001004b86c4cb78549b34bab61a3b1800e23bfeb5b3ec390074041536a7e3cbd97f5f04cf0f857155a8928eaa29ebfd11cfbbad3ba70efea7bda3226c6a8d370a4cd303f714486b6ebc225985a638471e6ef571cc92a4613c00b8fa65d61ccee0cbe5f36330c9a01f4183559f1bef24cc2917c6d913e3a541333a1d05d9bed22b38cb")]
```
Comment (stephentoub) [inline src/System.Net.Http.Json/src/Properties/InternalsVisibleTo.cs]: We try very hard not to use InternalsVisibleTo; we've stamped them out in most places, and I would like not to add more. .  There are a variety of reasons for this but, for example, this will prevent trimming anything in the assembly.  Can we please remove this?

File: src/System.Net.Http.Json/src/System/Net/Http/Json/HttpContentJsonExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Net.Http.Json
+{
+    public static class HttpContentJsonExtensions
+    {
+        public static Task<object?> ReadFromJsonAsync(this HttpContent content, Type type, JsonSerializerOptions? options = null, CancellationToken cancellationToken = default)
+        {
+            return ReadFromJsonAsyncCore(content, type, options, cancellationToken);
```
Comment (stephentoub) [inline src/System.Net.Http.Json/src/System/Net/Http/Json/HttpContentJsonExtensions.cs]: Nit: elsewhere we've used expression-body members for one-liners, but here we're not?

File: src/System.Net.Http.Json/src/System/Net/Http/Json/HttpContentJsonExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+        public static Task<T> ReadFromJsonAsync<T>(this HttpContent content, JsonSerializerOptions? options = null, CancellationToken cancellationToken = default)
+        {
+            return ReadFromJsonAsyncCore<T>(content, options, cancellationToken);
+        }
+
+        private static async Task<object?> ReadFromJsonAsyncCore(HttpContent content, Type type, JsonSerializerOptions? options, CancellationToken cancellationToken)
+        {
+            Stream contentStream = await GetJsonStreamFromContentAsync(content).ConfigureAwait(false);
+            return await JsonSerializer.DeserializeAsync(contentStream, type, options, cancellationToken);
```
Comment (stephentoub) [inline src/System.Net.Http.Json/src/System/Net/Http/Json/HttpContentJsonExtensions.cs]: .ConfigureAwait(false)

File: src/System.Net.Http.Json/ref/System.Net.Http.Json.cs (hunk 1 above)
Comment (jozkee) [reply src/System.Net.Http.Json/ref/System.Net.Http.Json.cs]: That's not what we agreed on API review, but I do agree on the need to have an optional mediaType or an overload that does not take it.

File: src/System.Net.Http.Json/src/System/Net/Http/Json/HttpContentJsonExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+        private static async Task<T> ReadFromJsonAsyncCore<T>(HttpContent content, JsonSerializerOptions? options, CancellationToken cancellationToken)
+        {
+            Stream contentStream = await GetJsonStreamFromContentAsync(content).ConfigureAwait(false);
+            return await JsonSerializer.DeserializeAsync<T>(contentStream, options, cancellationToken).ConfigureAwait(false);
+        }
+
+        private static async Task<Stream> GetJsonStreamFromContentAsync(HttpContent content)
+        {
+            ValidateMediaType(content.Headers.ContentType?.MediaType);
+            Debug.Assert(content.Headers.ContentType != null);
```
Comment (stephentoub) [inline src/System.Net.Http.Json/src/System/Net/Http/Json/HttpContentJsonExtensions.cs]: If we're able to assert that ContentType != null, why do we use the null-coalescing operator on the previous line?  And is this assert actually valid... what guarantees ContentType is non-null?

File: src/System.Net.Http.Json/src/System/Net/Http/Json/HttpContentJsonExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+                jsonStream = new TranscodingReadStream(jsonStream, sourceEncoding);
+            }
+
+            return jsonStream;
+        }
+
+        private static void ValidateMediaType(string? mediaType)
+        {
+            if (mediaType != JsonContent.JsonMediaType &&
+                 mediaType != MediaTypeNames.Text.Plain)
```
Comment (stephentoub) [inline src/System.Net.Http.Json/src/System/Net/Http/Json/HttpContentJsonExtensions.cs]: Nit: indentation

File: src/System.Net.Http.Json/src/System/Net/Http/Json/HttpContentJsonExtensions.cs (hunk 5 above)
Comment (jozkee) [reply src/System.Net.Http.Json/src/System/Net/Http/Json/HttpContentJsonExtensions.cs]: `ValidateMediaType` throws if the mediaType argument is null, it will be null if ContentType is null, it will be "application/json" or "text/plain" if is not null.

File: src/System.Net.Http.Json/src/System/Net/Http/Json/JsonContent.cs [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+        internal const string JsonMediaType = "application/json";
+        internal static readonly MediaTypeHeaderValue DefaultMediaType = MediaTypeHeaderValue.Parse(string.Format("{0} {1}", JsonMediaType, Encoding.UTF8.WebName));
+
+        private readonly JsonSerializerOptions? _jsonSerializerOptions;
+        public Type ObjectType { get; }
+        public object? Value { get; }
+
+        private JsonContent(object? value, Type inputType, MediaTypeHeaderValue mediaType, JsonSerializerOptions? options)
+        {
+            if (mediaType == null)
```
Comment (stephentoub) [inline src/System.Net.Http.Json/src/System/Net/Http/Json/JsonContent.cs]: I thought we were allowing mediaType to be null, and actually making it an optional parameter (defaulting to null) to avoid needing additional Create overloads that didn't take it.

Comment (jozkee) [reply src/System.Net.Http.Json/src/System/Net/Http/Json/JsonContent.cs]: In that case we need to mark it as such (`MediaTypeHeaderValue?`).

Comment (stephentoub) [reply src/System.Net.Http.Json/src/System/Net/Http/Json/JsonContent.cs]: Yes, that was the decision in API review as far as I remember: `MediaTypeHeaderValue? mediaType = null`.

File: src/System.Net.Http.Json/src/System/Net/Http/Json/HttpContentJsonExtensions.cs (hunk 5 above)
Comment (stephentoub) [reply src/System.Net.Http.Json/src/System/Net/Http/Json/HttpContentJsonExtensions.cs]: Ok

File: src/System.Net.Http.Json/ref/System.Net.Http.Json.cs (hunk 1 above)
Comment (jozkee) [reply src/System.Net.Http.Json/ref/System.Net.Http.Json.cs]: Or maybe it is just missing in the spec.
https://github.com/dotnet/runtime/issues/33566#issuecomment-598845161

**Comment by joperezr** [comment]
@ericstj can you please take a look to aab639f1 to make sure that the suppression added and the package index is what you expect?

File: src/System.Net.Http.Json/src/Properties/InternalsVisibleTo.cs (hunk 2 above)
Comment (jozkee) [reply src/System.Net.Http.Json/src/Properties/InternalsVisibleTo.cs]: Sure, I can workaround this. I just added it since the original code on aspnetcore is being tested this way.

File: src/System.Net.Http.Json/ref/System.Net.Http.Json.cs (hunk 1 above)
Comment (jozkee) [reply src/System.Net.Http.Json/ref/System.Net.Http.Json.cs]: Agree.

File: src/System.Net.Http.Json/src/System/Net/Http/Json/JsonContent.cs [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+
+        private Stream GetStreamToWriteTo(Stream targetStream)
+        {
+            Stream jsonStream = targetStream;
+            Encoding? targetEncoding = GetEncoding(Headers.ContentType.CharSet);
+
+            // Wrap provided stream into a transcoding stream that buffers the data transcoded from utf-8 to the targetEncoding.
+            if (targetEncoding != null && targetEncoding != Encoding.UTF8)
+            {
+                jsonStream = new TranscodingWriteStream(jsonStream, targetEncoding);
```
Comment (stephentoub) [inline src/System.Net.Http.Json/src/System/Net/Http/Json/JsonContent.cs]: This TranscodingWriteStream is renting arrays from the pool, and those arrays are then returned as part of its Dispose.  But the returned Stream isn't being disposed in SerializeToStreamAsync, which means we're taking buffers out of the pool and never returning them.

**Comment by joperezr** [comment]
Adding No Merge label to wait till the runtime PR has been approved and checked in before we merge this one.

File: src/System.Net.Http.Json/ref/System.Net.Http.Json.cs [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+// See the LICENSE file in the project root for more information.
+// ------------------------------------------------------------------------------
+// Changes to this file must follow the https://aka.ms/api-review process.
+// ------------------------------------------------------------------------------
+
+namespace System.Net.Http.Json
+{
+    public static partial class HttpClientJsonExtensions
+    {
+        public static System.Threading.Tasks.Task<object?> GetFromJsonAsync(this System.Net.Http.HttpClient client, string? requestUri, System.Type type, System.Text.Json.JsonSerializerOptions? options = null, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) { throw null; }
```
Comment (jozkee) [inline src/System.Net.Http.Json/ref/System.Net.Http.Json.cs]: @stephentoub Should we call it `Type returnType` instead? here and for `HttpContentJsonExtensions.ReadFromJsonAsync` since that's the name of the param in JsonSerializer.Deserialize.

https://docs.microsoft.com/en-us/dotnet/api/system.text.json.jsonserializer.deserialize?view=netcore-3.1#System_Text_Json_JsonSerializer_Deserialize_System_String_System_Type_System_Text_Json_JsonSerializerOptions_

File: src/System.Net.Http.Json/ref/System.Net.Http.Json.cs (hunk 1 above)
Comment (stephentoub) [reply src/System.Net.Http.Json/ref/System.Net.Http.Json.cs]: > That's not what we agreed on API review

https://youtu.be/_AHbjIS8_0I?t=4272

**Comment by ericstj** [comment]
Why is code-review on content and test fixing happening in release?  Is this specific to release code?  If not, polish this and finish in master first.  Release PR should be cherry-pick and whatever minor changes need to be done to port.  I wouldn't expect tons of feedback here, it is not efficient to do that in two places.

File: src/System.Net.Http.Json/src/System/Net/Http/Json/TranscodingReadStream.cs [hunk 10]
```diff
@@ -0,0 +1,244 @@
...
+                _disposed = true;
+
+                Debug.Assert(_charBuffer.Array != null);
+                ArrayPool<char>.Shared.Return(_charBuffer.Array);
+
+                Debug.Assert(_byteBuffer.Array != null);
+                ArrayPool<byte>.Shared.Return(_byteBuffer.Array);
+
+                Debug.Assert(_overflowBuffer.Array != null);
+                ArrayPool<byte>.Shared.Return(_overflowBuffer.Array);
```
Comment (stephentoub) [inline src/System.Net.Http.Json/src/System/Net/Http/Json/TranscodingReadStream.cs:237]: This should null out the various buffer fields.  Otherwise, misuse of this stream could end up accessing arrays that have already been returned to the pool.

**Comment by stephentoub** [comment]
> Why is code-review on content and test fixing happening in release?

Simply because this is the PR I heard you refer to in the meeting just now / you sent out a link to.  I didn't receive any notifications that the other PR in dotnet/runtime had been updated, and saw this as being newer than the last time I looked at dotnet/runtime.

**Comment by jozkee** [comment]
@stephentoub for the record, the PR in dotnet/runtime already contains the set of changes that you are reviewing here.

**Comment by stephentoub** [comment]
Ok.  I didn't receive a notification that it had been updated.  I only received a link from Eric to this PR.  Hence my commenting here.

I'm halfway through.  How would you like me to proceed?

**Comment by jozkee** [comment]
Just to keep PR comments centralized, you can continue on https://github.com/dotnet/runtime/pull/33459 

I will address the feedback on both PRs.

File: pkg/Microsoft.Private.CoreFx.NETCoreApp/Microsoft.Private.CoreFx.NETCoreApp.pkgproj [hunk 11]
```diff
@@ -60,6 +60,14 @@
             System.ComponentModel.Composition;
         </Value>
     </ValidatePackageSuppression>
+    <!-- Permit inbox revsion of assemblies that are part of shared framework but also ship as packages. This will
+    make sure other packages that have a reference to them will carry a package reference to the latest serviced
+    version. -->
+    <ValidatePackageSuppression Include="PermitInbox">
```
Comment (ericstj) [inline pkg/Microsoft.Private.CoreFx.NETCoreApp/Microsoft.Private.CoreFx.NETCoreApp.pkgproj:66]: Why was this required? Can you please share the error message you were getting?

Comment (joperezr) [reply pkg/Microsoft.Private.CoreFx.NETCoreApp/Microsoft.Private.CoreFx.NETCoreApp.pkgproj:66]: The private transport package was failing due to a version inconsistency, System.Text.Json was expected to be inbox with version 4.0.1.0 but was found to have 4.0.1.1. Here is the error message I got:

```
C:\Users\joperezr\.nuget\packages\microsoft.dotnet.build.tasks.packaging\1.0.0-beta.20113.5\build\Packaging.targets(1161,5): error : File System.Text.Json, version 4.0.1.1 was included framework package Microsoft.Private.CoreFx.NETCoreApp/4.7.0-dev.20167.1 but that version is not considered inbox in package index F:\git\corefx\pkg\Microsoft.Private.PackageBaseline\packageIndex.json.  Please add it with appropriate InboxOn entry for netcoreapp3.1 or suppress this message with PermitInbox suppression. [F:\git\corefx\pkg\Microsoft.Private.CoreFx.NETCoreApp\Microsoft.Private.CoreFx.NETCoreApp.pkgproj]
```

File: src/System.Net.Http.Json/tests/FunctionalTests/TranscodingReadStreamTests.cs [hunk 12]
```diff
@@ -0,0 +1,273 @@
...
+
+        internal static TheoryData GetUnicodeText(int maxCharBufferSize)
+        {
+            return new TheoryData<string>
+            {
+                new string('Æ', count: 7),
+                new string('A', count: maxCharBufferSize - 1) + 'Æ',
+                "AbĀāĂăĄąĆŊŋŌōŎŏŐőŒœŔŕŖŗŘřŚşŠšŢţŤťŦŧŨũŪūŬŭŮůŰűŲųŴŵŶŷŸŹźŻżŽžſAbc",
+               "Abcஐஒஓஔகஙசஜஞடணதநனபமயரறலளழவஷஸஹ",
+               "☀☁☂☃☄★☆☇☈☉☊☋☌☍☎☏☐☑☒☓☚☛☜☝☞☟☠☡☢☣☤☥☦☧☨☩☪☫☬☭☮☯☰☱☲☳☴☵☶☷☸",
```
Comment (danmoseley) [inline src/System.Net.Http.Json/tests/FunctionalTests/TranscodingReadStreamTests.cs]: Please convert these to `\uxxxx` form so that the file can be saved as ASCII. We required that in this repo as historically files have gotten garbled otherwise. Hopfeully there's an easy way to convert them..

**Comment by joperezr** [comment]
Closing this PR in favor of #42889 which is targeting the new branch that was created for this project.

File: src/System.Net.Http.Json/tests/FunctionalTests/TranscodingReadStreamTests.cs (hunk 12 above)
Comment (jozkee) [reply src/System.Net.Http.Json/tests/FunctionalTests/TranscodingReadStreamTests.cs]: Done in master and in blazor branch. 
Thanks.

Comment (danmoseley) [reply src/System.Net.Http.Json/tests/FunctionalTests/TranscodingReadStreamTests.cs]: Thanks, guaranteed you would have gotten unit tests failing on some random day otherwise.
